Sources/MapboxCoreNavigation/MapboxRoutingProvider.swift

```diff
@@ -0,0 +1,338 @@
+     Valid only for the same instance of `MapboxRoutingProvider` that issued it.
+     */
+    public typealias RequestId = UInt64
+    
+    /**
+     A request handler for the ongoing router action.
+     
+     You can use this instance to cancel ongoing task if needed. Retaining this handler will keep related `MapboxRoutingProvider` from deallocating.
+     */
+    public struct Request: NavigationProviderRequest {
+        /**
+         Related request identifier.
+         */
+        public let requestIdentifier: RequestId
+        
+        // Intended retain cycle to prevent deallocating. `Request` will be deleted once request completes.
+        let router: MapboxRoutingProvider
+        
+        /**
+         Cancels the request if it is still active.
+         */
+        public func cancel() {
+            router.router.cancelRequest(forToken: requestIdentifier)
+        }
+    }
+    
+    /**
+     Defines source of routing engine to be used for requests.
+     */
+    public enum Source {
+        /**
+         Fetch data online only
+         
+         Such `MapboxRoutingProvider` is equivalent of using bare `Directions` wrapper.
+         */
+        case online
+        /**
+         Use online data only
+         
+         In order for such `MapboxRoutingProvider` to function properly, proper navigation data should be available onboard. `.offline` router will not be able to refresh routes.
+         */
+        case offline
+        /**
+         Attempts to use `online` with fallback to `offline`.
+         
+         `.hybrid` router will be able to refresh routes only using internet connection.
+         */
+        case hybrid
+        
+        var nativeSource: RouterType {
+            switch self {
+            case .online:
+                return .online
+            case .offline:
+                return .onboard
+            case .hybrid:
+                return .hybrid
+            }
+        }
+    }
+    
+    static var __testRoutesStub: ((_: RouteOptions, _: @escaping Directions.RouteCompletionHandler) -> Request?)? = nil
+    
+    // MARK: - Properties
+    /**
+     List of ongoing tasks for the router.
+     
+     You can see if router is busy with somethin, or used related `Request.cancel()` to cancel requests as needed.
+     */
+    public private(set) var activeRequests: [RequestId : Request] = [:]
+    /**
+     Configured routing engine source.
+     */
+    public let source: Source
+    
+    private let requestsLock = NSLock()
+    private let router: MapboxNavigationNative_Private.RouterInterface
+    private let settings: NavigationSettings
+    
+    //MARK: - Initialization
+    
+    /**
+     Initializes new `MapboxRoutingProvider`.
+     
+     - parameter source: routing engine source to use.
+     - parameter settings: settings object, used to get credentials and cache configuration.
+     */
+    public init(_ source: Source = .hybrid, settings: NavigationSettings = .shared) {
```

Comment:
Settings are expected as a parameter to not introduce implicit dependency on it. Plus it is an easy option to enable more flexibility in case it's needed.